### PR TITLE
 make .nunique callable from inside the square brackets

### PR DESCRIPTION
**Pull request by pradkrish**
WIP for #2999

### CODE REVIEW

**Comment by oleksiyskononenko** [comment]
@pradkrish Thanks for PR. You should merge `main` onto your branch to resolve conflicts appearing after https://github.com/h2oai/datatable/pull/3024 merge.

File: tests/dt/test-nunique.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#-------------------------------------------------------------------------------
+#  nunique
+#-------------------------------------------------------------------------------
+
+@pytest.mark.parametrize("src", srcs_all)
+def test_dt_nunique(src):
+    df = dt.Frame(src)
+    EXP = df.nunique()
+    RES = df[:, dt.nunique(f[:])]
+    assert_equals(EXP,RES)
```
Comment (oleksiyskononenko) [inline tests/dt/test-nunique.py:68]: Can we add a test to make sure `nunique()` works correctly in the presence of a group by clause? Something similar to what you did for `countna()`.

**Comment by oleksiyskononenko** [comment]
When I compile this PR, I'm getting quite a lot of warnings. Could you please fix them, so that we have a clean build output?

File: docs/api/dt/nunique.rst [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    cols: Expr
+        Input columns.
+
+    return: Expr
+        f-expression having one row, and the same names and number of columns
+        as in `cols`. All the returned column stypes are `int64`.
+
+    except: TypeError
+        The exception is raised when one of the columns from `cols`
+        has an obj64 type.
```
[the remark this answers is not in the document]
Comment (pradkrish) [reply docs/api/dt/nunique.rst]: Done

File: docs/api/dt/nunique.rst [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        -- + -----  -----  -----
+         0 |     1     NA      1
+         1 |     1      2      2
+         2 |     2      3      1
+         3 |    NA      4      1
+         4 |     1     NA      2
+         5 |     2      5      4
+        [6 rows x 3 columns]
+
+    Count the number of unique values of all rows:
```
[the remark this answers is not in the document]
Comment (pradkrish) [reply docs/api/dt/nunique.rst]: Done

File: docs/api/dt/nunique.rst [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        >>> df[:, dt.nunique(f[:])]
+           |     A      B      C
+           | int64  int64  int64
+        -- + -----  -----  -----
+         0 |     2      4      3
+        [1 row x 3 columns]
+
+    Count the number of unique values of column `B`:
```
[the remark this answers is not in the document]
Comment (pradkrish) [reply docs/api/dt/nunique.rst]: Done

File: docs/api/index-api.rst [hunk 5]
```diff
@@ -168,7 +168,9 @@ Functions
     * - :func:`median()`
       - Find the median element per column
     * - :func:`min()`
-      - Find the smallest element per column
+      - Find the smallest element per a column
```
[the remark this answers is not in the document]
Comment (pradkrish) [reply docs/api/index-api.rst]: That was definitely by accident. Glad to hear google results confirm that :) Done.

File: tests/dt/test-nunique.py (hunk 1 above)
Comment (pradkrish) [reply tests/dt/test-nunique.py:68]: Good point. It wasn't working in the presence of `by()`. Added an extra logic and a test to that end.

**Comment by pradkrish** [comment]
> When I compile this PR, I'm getting quite a lot of warnings. Could you please fix them, so that we have a clean build output?

I didn't see so many warnings at my end. Kindly check if that's still the case. Thanks.

**Comment by oleksiyskononenko** [comment]
@pradkrish Thanks, regarding the warnings please take a look here https://ci.appveyor.com/project/h2oops/datatable/builds/40912944/job/06rbeki8lhu4y4w5?fullLog=true#L398

**Comment by pradkrish** [comment]
> @pradkrish Thanks, regarding the warnings please take a look here https://ci.appveyor.com/project/h2oops/datatable/builds/40912944/job/06rbeki8lhu4y4w5?fullLog=true#L398

@oleksiyskononenko Do check now, it should be fixed. I checked on a Mac system and those compiler warnings has disappeared. It's also ready for review.

File: src/core/expr/head_reduce_unary.cc [hunk 6]
```diff
@@ -749,8 +751,87 @@ static Column compute_max(Column&& arg, const Groupby& gby) {
...
+//------------------------------------------------------------------------------
+// nunique(A:grouped)
+//------------------------------------------------------------------------------
+
+// T is the type of the input column
+
+
+template <typename T, bool NA>
+static Column _gnunique(Column&& arg, const Groupby& gby) {
+  return Column(new CountGrouped_ColumnImpl<T,NA>(std::move(arg), gby));
```
Comment (oleksiyskononenko) [inline src/core/expr/head_reduce_unary.cc:763]: Are you sure you can simply reuse `CountGrouped_ColumnImpl ` for `nunique`? Consider this example
```
>>>DT = dt.Frame([1, 1, 2])
>>>DT[:, {"nunique" : dt.nunique(f[0])}, dt.by(f[0])]
```
the result is the same as for `dt.count()`:
```
   |    C0  nunique
   | int32    int64
-- + -----  -------
 0 |     1        2
 1 |     2        1
[2 rows x 2 columns]
```
I would expect it returning
```
   |    C0  nunique
   | int32    int64
-- + -----  -------
 0 |     1        1
 1 |     2        1
[2 rows x 2 columns]
```

Comment (pradkrish) [reply src/core/expr/head_reduce_unary.cc:763]: Thanks @oleksiyskononenko, you are absolutely right, a massive oversight on my part. I will try to fix it.

File: src/core/expr/head_reduce_unary.cc [hunk 7]
```diff
@@ -749,8 +751,125 @@ static Column compute_max(Column&& arg, const Groupby& gby) {
...
+    bool get_element(size_t i, int64_t* out) const override {
+      T value;
+      bool isvalid = arg.get_element(i, &value);
+      if (isvalid) {
+        size_t i0, i1;
+        groupby.get_group(i, &i0, &i1);
+        *out = 1;
+      } else {
+        *out = 0;
+      }
```
Comment (oleksiyskononenko) [inline src/core/expr/head_reduce_unary.cc:785]:
```suggestion
      *out = arg.get_element(i, &value);
```

**Review by oleksiyskononenko — APPROVED** [review]

**Comment by oleksiyskononenko** [comment]
Thanks @pradkrish, looks awesome, I just made some minor cosmetic corrections here and there.

**Pull request merged**